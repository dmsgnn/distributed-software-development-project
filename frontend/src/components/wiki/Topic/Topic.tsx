import { Box, HStack, Text } from '@chakra-ui/react';
import Image, { StaticImageData } from 'next/image';
import { CSSProperties } from 'react';
import Deployment from '@/images/deployment.png';
import Design from '@/images/design.png';
import Requirements from '@/images/requirements.png';
import Testing from '@/images/testing.png';
import Development from '@/images/development.png';

export enum Topics {
	Requirements,
	Design,
	Development,
	Testing,
	Deployment,
}

interface TopicProps {
	variant: Topics;
}

const Topic: React.FC<TopicProps> = ({ variant }) => {
	const { img, bg, title, url } = Variants[variant];

	return (
		<HStack as="a" sx={styles} bg={bg} href={url}>
			<Text>{title}</Text>
			<Image src={img} alt="" height={100} width={150} />
		</HStack>
	);
};

const styles = {
	alignItems: 'center',
	justifyContent: 'space-between',
	borderRadius: 25,
	padding: '10px 16px 10px 32px',
	width: 341,
	height: 142,
	color: 'white',
	boxSizing: 'border-box',
} as CSSProperties;

const Variants: { img: StaticImageData; bg: string; title: string; url: string }[] = [
	{
		img: Requirements,
		bg: '#4042E2',
		title: 'Requirements',
		url: '/wiki/requirements',
	},
	{
		img: Design,
		bg: '#FF8077',
		title: 'Design',
		url: '/wiki/requirements',
	},
	{
		img: Development,
		bg: '#7D97F4',
		title: 'Development',
		url: '/wiki/requirements',
	},
	{
		img: Testing,
		bg: '#A13976',
		title: 'Testing',
		url: '/wiki/requirements',
	},
	{
		img: Deployment,
		bg: '#F29A33',
		title: 'Deployment',
		url: '/wiki/requirements',
	},
];

export default Topic;
