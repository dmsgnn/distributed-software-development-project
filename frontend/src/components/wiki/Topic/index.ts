export { default, Topics } from './Topic';
