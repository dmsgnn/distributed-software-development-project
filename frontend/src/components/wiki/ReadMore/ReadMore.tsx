import { Flex, Heading } from '@chakra-ui/react';
import React from 'react';
import Topic, { Topics } from '../Topic';

interface ReadMoreProps {
	activeTopic?: Topics;
}

const ReadMore: React.FC<ReadMoreProps> = ({ activeTopic }) => {
	const topics = [Topics.Requirements, Topics.Design, Topics.Deployment, Topics.Testing, Topics.Deployment];

	return (
		<>
			<Heading as="h3" fontSize="24px">
				Read more
			</Heading>
			<Flex wrap="wrap" gap="2rem" justifyContent="center" mt="3rem">
				{activeTopic ? <Topic variant={activeTopic} /> : null}
				<Topic variant={Topics.Design} />
				<Topic variant={Topics.Development} />
				<Topic variant={Topics.Testing} />
				<Topic variant={Topics.Deployment} />
			</Flex>
		</>
	);
};

export default ReadMore;
