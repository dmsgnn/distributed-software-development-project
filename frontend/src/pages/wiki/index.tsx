import { MainLayout } from '@/components/MainLayout/MainLayout';
import { MainNavigation } from '@/components/MainNavigation/MainNavigation';
import Topic, { Topics } from '@/components/wiki/Topic';
import { Container, Flex, Heading, Text } from '@chakra-ui/react';
import { InferGetStaticPropsType, NextPage } from 'next';
import { serverSideTranslations } from 'next-i18next/serverSideTranslations';

type WikiPageProps = InferGetStaticPropsType<typeof getStaticProps>;

const Wiki: NextPage<WikiPageProps> = () => {
	return (
		<MainLayout navigation={<MainNavigation />}>
			<Container maxW="container.xl" mt={16}>
				<Heading as="h1" textAlign="center" className="albra" mb="0.5em">
					SSLDC
				</Heading>
				<Text mx="auto" w="500px" fontSize="1rem" mb="5rem">
					The point of DSEC is to help you implement a SSDLC. Therefore, we have carefully selected some tips for every
					stage of the process.
				</Text>

				<Flex wrap="wrap" gap="2rem" justifyContent="center" mt="3rem">
					<Topic variant={Topics.Requirements} />
					<Topic variant={Topics.Design} />
					<Topic variant={Topics.Development} />
					<Topic variant={Topics.Testing} />
					<Topic variant={Topics.Deployment} />
				</Flex>
			</Container>
		</MainLayout>
	);
};

export const getStaticProps = async ({ locale }) => {
	return {
		props: {
			...(await serverSideTranslations(String(locale), ['common'])),
		},
	};
};

export default Wiki;
