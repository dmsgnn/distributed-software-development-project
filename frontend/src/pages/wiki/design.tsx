import { Topics } from '@/components/wiki/Topic';
import WikiPage from '@/components/wiki/WikiPage';
import { serverSideTranslations } from 'next-i18next/serverSideTranslations';
import React from 'react';

const DesignWikiPage = () => {
	const content = [
		{
			title: 'Conduct a threat modeling exercise',
			description:
				'Identify the potential threats that the software may be exposed to and determine how they will be mitigated.',
		},
		{
			title: 'Design for security',
			description:
				'Consider security at every step of the design process and ensure that the software is designed to be secure.',
		},
		{
			title: 'Use secure design patterns',
			description:
				'Implement security controls (e.g., authentication, authorization, encryption) to protect the software from threats.Leverage design patterns that have been proven to be secure, such as the "principle of least privilege" and "fail-safe defaults".',
		},
		{
			title: 'Use a design review',
			description: 'Have a team of developers or security experts review the design to ensure that it is secure.',
		},
	];

	return <WikiPage title={'Design'} content={content} topic={Topics.Design} />;
};

export const getStaticProps = async ({ locale }) => {
	return {
		props: {
			...(await serverSideTranslations(String(locale), ['common'])),
		},
	};
};

export default DesignWikiPage;
