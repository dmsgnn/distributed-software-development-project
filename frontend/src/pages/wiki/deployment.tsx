import ReadMore from '@/components/wiki/ReadMore';
import Tip from '@/components/wiki/Tip';
import { Topics } from '@/components/wiki/Topic';
import WikiPage from '@/components/wiki/WikiPage';
import { serverSideTranslations } from 'next-i18next/serverSideTranslations';
import React from 'react';

const DeploymentWikiPage = () => {
	const accentColor = '#F29A33';

	return (
		<WikiPage title="Deployment" accentColor={accentColor}>
			<Tip
				title="Monitor and update security controls"
				description="Regularly monitor the software for security vulnerabilities and update security controls as needed to keep the software secure."
				align={'left'}
				index={1}
			/>
			<Tip
				title="Keep your secrets a secret"
				description="Be careful where you choose to store your secrets and never make them publicly available. If you do, change them immediately! The DSEC tool can help you check if you have any exposed secrets in your code."
				align={'right'}
				index={2}
				accentColor={accentColor}
			/>
			<Tip
				title="Use secure updates"
				description="Use secure methods for updating the software, such as signed update packages and secure communications channels."
				align={'left'}
				index={3}
			/>
			<Tip
				title="Monitor for security breaches"
				description="Use monitoring tools to detect and alert on potential security breaches."
				align={'right'}
				index={4}
				accentColor={accentColor}
			/>
			<Tip
				title="Respond to security incidents"
				description="Have a plan in place for responding to security incidents, including steps for containing the incident, mitigating the impact, and restoring the software to a secure state."
				align={'left'}
				index={5}
			/>
			<ReadMore activeTopic={Topics.Deployment} />
		</WikiPage>
	);
};

export const getStaticProps = async ({ locale }) => {
	return {
		props: {
			...(await serverSideTranslations(String(locale), ['common'])),
		},
	};
};

export default DeploymentWikiPage;
