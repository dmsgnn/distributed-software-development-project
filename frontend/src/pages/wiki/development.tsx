import ReadMore from '@/components/wiki/ReadMore';
import Tip from '@/components/wiki/Tip';
import { Topics } from '@/components/wiki/Topic';
import WikiPage from '@/components/wiki/WikiPage';
import { serverSideTranslations } from 'next-i18next/serverSideTranslations';
import React from 'react';

const DevelopmentWikiPage = () => {
	const accentColor = '#7D97F4';

	return (
		<WikiPage title="Development" accentColor={accentColor}>
			<Tip
				title="Implement security controls"
				description="Incorporate security measures into the software to protect against identified threats. This can include measures such as input validation, authentication and authorization, and encryption."
				align={'left'}
				index={1}
			/>
			<Tip
				title="Use secure coding practices"
				description="Follow best practices for writing secure code, such as properly handling errors and exceptions, avoiding the use of hard-coded passwords, and sanitizing user input."
				align={'right'}
				index={2}
				accentColor={accentColor}
			/>
			<Tip
				title="Use a static code analyzer"
				description="Use a tool that automatically checks the code for vulnerabilities and security issues. The DSEC tool can help you set this up for various programming languages."
				align={'left'}
				index={3}
			/>
			<Tip
				align={'right'}
				index={4}
				accentColor={accentColor}
				title="Use libraries and frameworks"
				description="Leverage existing libraries and frameworks that provide built-in security features, such as input validation and output encoding."
			/>
			<Tip
				title="Use a secure development environment"
				description="Use a development environment that is configured with secure settings and tools to help prevent security breaches."
				align={'left'}
				index={5}
			/>
			<ReadMore activeTopic={Topics.Deployment} />
		</WikiPage>
	);
};

export const getStaticProps = async ({ locale }) => {
	return {
		props: {
			...(await serverSideTranslations(String(locale), ['common'])),
		},
	};
};

export default DevelopmentWikiPage;
