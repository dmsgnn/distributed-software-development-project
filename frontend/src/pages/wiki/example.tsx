import { MainLayout } from '@/components/MainLayout/MainLayout';
import { MainNavigation } from '@/components/MainNavigation/MainNavigation';
import ReadMore from '@/components/wiki/ReadMore';
import Topic, { Topics } from '@/components/wiki/Topic';
import { Container, Heading, Flex, Text } from '@chakra-ui/react';
import { InferGetStaticPropsType, NextPage } from 'next';
import { serverSideTranslations } from 'next-i18next/serverSideTranslations';

type WikiPageProps = InferGetStaticPropsType<typeof getStaticProps>;

const Wiki: NextPage<WikiPageProps> = () => {
	return (
		<MainLayout navigation={<MainNavigation />}>
			<Container maxW="800px" mt={16}>
				<Heading as="h2">Requirements</Heading>
				<br />
				<Text>
					Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer id ullamcorper augue. Pellentesque suscipit
					sem nibh, et ullamcorper arcu sollicitudin in. Curabitur mollis purus eget commodo pharetra. Sed gravida, orci
					eget eleifend tempor, ante enim scelerisque massa, non ullamcorper dolor elit id sem. Donec non mauris
					sodales, rhoncus ipsum non, consequat nibh. Quisque tincidunt iaculis nunc, quis varius ipsum auctor volutpat.
				</Text>
				<br />
				<br />
				<Heading as="h3" fontSize={24}>
					Aliquam congue, turpis quis iaculis
				</Heading>
				<br />
				<Text>
					Augue quam semper quam, vel hendrerit tortor tellus nec ipsum. Nullam urna ante, tempor commodo libero sed,
					aliquam molestie turpis. Nam facilisis tellus ac velit ultricies aliquet. Donec luctus mi eget metus
					venenatis, a ullamcorper nulla ornare. Proin a turpis nec magna venenatis elementum eget sit amet elit. Cras
					auctor a augue sed facilisis. Aenean semper metus urna, sit amet feugiat magna sodales varius. Nullam
					dignissim elit eget nisl euismod malesuada. Vivamus ornare elit eu nibh interdum iaculis ut nec urna
				</Text>

				<br />
				<br />
				<ReadMore activeTopic={Topics.Requirements} />
			</Container>
		</MainLayout>
	);
};

export const getStaticProps = async ({ locale }) => {
	return {
		props: {
			...(await serverSideTranslations(String(locale), ['common'])),
		},
	};
};

export default Wiki;
