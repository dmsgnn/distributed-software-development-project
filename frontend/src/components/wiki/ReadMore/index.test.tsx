import { Topics } from '@/components/wiki/Topic';
import { axe } from 'jest-axe';

import { act, render } from 'test-utils';
import ReadMore from './';

const mockRouterPush = jest.fn();
jest.mock('next/router', () => ({ useRouter: jest.fn(() => ({ push: mockRouterPush })) }));

describe('ReadMore', () => {
	it('should matches snapshot', () => {
		const { asFragment } = render(<ReadMore activeTopic={Topics.Design} />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('should be accessible', async () => {
		const { container } = render(<ReadMore activeTopic={Topics.Design} />);

		let results: Awaited<ReturnType<typeof axe>>;
		await act(async () => {
			results = await axe(container);
		});

		expect(results).toHaveNoViolations();
	});
});
