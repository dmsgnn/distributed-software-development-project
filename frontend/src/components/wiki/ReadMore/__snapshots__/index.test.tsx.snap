// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReadMore should matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1qsf17s"
  >
    <h3
      class="chakra-heading css-16iqw5x"
    >
      Read more
    </h3>
    <div
      class="css-v6gzm5"
    >
      <a
        class="chakra-stack css-j683cg"
        href="/wiki/design"
      >
        <p
          class="chakra-text css-0"
        >
          Design
        </p>
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=384&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=256&q=75 1x, /_next/image?url=%2Fimg.jpg&w=384&q=75 2x"
          style="color: transparent;"
          width="150"
        />
      </a>
      <a
        class="chakra-stack css-j683cg"
        href="/wiki/design"
      >
        <p
          class="chakra-text css-0"
        >
          Design
        </p>
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=384&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=256&q=75 1x, /_next/image?url=%2Fimg.jpg&w=384&q=75 2x"
          style="color: transparent;"
          width="150"
        />
      </a>
      <a
        class="chakra-stack css-n2ydgl"
        href="/wiki/development"
      >
        <p
          class="chakra-text css-0"
        >
          Development
        </p>
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=384&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=256&q=75 1x, /_next/image?url=%2Fimg.jpg&w=384&q=75 2x"
          style="color: transparent;"
          width="150"
        />
      </a>
      <a
        class="chakra-stack css-1c6oe1o"
        href="/wiki/testing"
      >
        <p
          class="chakra-text css-0"
        >
          Testing
        </p>
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=384&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=256&q=75 1x, /_next/image?url=%2Fimg.jpg&w=384&q=75 2x"
          style="color: transparent;"
          width="150"
        />
      </a>
      <a
        class="chakra-stack css-1qnjf86"
        href="/wiki/deployment"
      >
        <p
          class="chakra-text css-0"
        >
          Deployment
        </p>
        <img
          alt=""
          data-nimg="1"
          decoding="async"
          height="100"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=384&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=256&q=75 1x, /_next/image?url=%2Fimg.jpg&w=384&q=75 2x"
          style="color: transparent;"
          width="150"
        />
      </a>
    </div>
  </div>
</DocumentFragment>
`;
